### PR TITLE
Generic Framework : Repetitive testing with different set of data. -WAR-2298

### PR DESCRIPTION
**Pull request by ryadaval-warrior**

### CODE REVIEW

**Comment by ryadaval-warrior** [comment]
use these commands to test :
-------------------------------------------------
genericdatafile path should be relative to testcase pah or absolute dir.
1) Warrior wftests/warrior_tests/testcases/framework_tests/test_log_messages_gen.xml -genericdatafile data_generic.xls -gen_no_of_samples 2
2) Warrior wftests/warrior_tests/testcases/framework_tests/test_log_messages_gen.xml -gen_no_of_samples 2
3) Warrior wftests/warrior_tests/testcases/framework_tests/test_log_messages_gen.xml -gen_no_of_samples 1 -gen_select_rows 0,1
4) Warrior wftests/warrior_tests/testcases/framework_tests/test_log_messages_gen.xml -gen_no_of_samples 2 -gen_shuffle_columns -gen_purge_db
5) Warrior wftests/warrior_tests/testcases/framework_tests/test_log_messages_gen.xml -gen_no_of_samples 2 -gen_shuffle_columns -gen_purge_db

File: .travis.yml [hunk 1]
```diff
@@ -35,6 +35,7 @@ install:
       pip3 install django==2.0.13;
       pip3 install djangorestframework==3.9.4;
       pip3 install xmltodict==0.12.0;
+      pip3 install pandas;
```
Comment (santhosh-warrior) [inline .travis.yml:38]: add the version pandas== 0.25.3

File: setup.py [hunk 2]
```diff
@@ -37,14 +37,13 @@
...
-                 'Programming Language :: Python :: 3.6',
-                 ],
+                 'Programming Language :: Python :: 3.6',],
     install_requires=["pexpect==4.8.0", "requests==2.21.0", "selenium==3.8.0",
                       "lxml==4.4.1", "paramiko==2.4.2", "pysnmp==4.4.9",
                       "pyvirtualdisplay==0.2.1", "kafka-python==1.4.6",
                       "cloudshell-automation-api==9.3.0.175525",
                       "pycryptodome==3.6.1", "Django==2.1.2", "xmltodict==0.12.0",
-                      "python-ldap", "django_auth_ldap",
+                      "python-ldap", "django_auth_ldap", "xlrd", "pandas",
```
Comment (santhosh-warrior) [inline setup.py:46]: add the package version for pandas and xlrd

File: warrior/WarriorCore/testcase_driver.py [hunk 3]
```diff
@@ -866,20 +875,198 @@ def check_robot_wrapper_case(testcase_filepath):
...
+                          " Please check data in given genericdatafile. Exiting!!")
+            exit(1)
+        gen_dict = {"gen_dict" : generic_data_dict}
+        Utils.data_Utils.update_datarepository(gen_dict)
+    return generic_data_dict
 
 def main(testcase_filepath, data_repository={}, tc_context='POSITIVE',
          runtype='SEQUENTIAL_KEYWORDS', tc_parallel=False, auto_defects=False, suite=None,
-         tc_onError_action=None, iter_ts_sys=None, queue=None, jiraproj=None):
+         tc_onError_action=None, iter_ts_sys=None, queue=None, jiraproj=None, jiraid=None):
```
Comment (santhosh-warrior) [inline warrior/WarriorCore/testcase_driver.py:1038]: Why Jiraid is added to the args

File: warrior/WarriorCore/Classes/war_cli_class.py [hunk 4]
```diff
@@ -346,6 +346,23 @@ def parser(cls, arglist):
         warrior_arg.add_argument('-encrypt', action='store', nargs='*',\
             dest="encrypt", help="encrypt data string")
 
+        warrior_arg.add_argument('-genericdatafile', action='store', nargs='?',\
+            dest="genericdatafile", help="generic data file for repetitive testing")
+
+        warrior_arg.add_argument('-gen_no_of_samples', action='store', nargs='?',\
```
Comment (santhosh-warrior) [inline warrior/WarriorCore/Classes/war_cli_class.py:352]: Use type=int or handle invalid inputs

File: warrior/WarriorCore/Classes/war_cli_class.py [hunk 5]
```diff
@@ -346,6 +346,23 @@ def parser(cls, arglist):
...
+            dest="gen_no_of_samples", help="number of random rows to \
+            select from generic data file")
+
+        warrior_arg.add_argument('-gen_select_rows', action='store', nargs='?',\
```
Comment (santhosh-warrior) [inline warrior/WarriorCore/Classes/war_cli_class.py:356]: Handle the invalid input scenario

File: warrior/WarriorCore/testcase_driver.py (hunk 3 above)
Comment (ryadaval-warrior) [reply warrior/WarriorCore/testcase_driver.py:870]: this is part of code change

File: warrior/WarriorCore/Classes/war_cli_class.py (hunk 5 above)
Comment (ryadaval-warrior) [reply warrior/WarriorCore/Classes/war_cli_class.py]: done

File: warrior/Framework/Utils/testrandom_utils.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    print_info("Total no. of combinations already tested : {}".format(len(records_in_db)))
+    print_info("Total no. of combinations remaining to test : {}".format(len(all_dict) - len(records_in_db)))
+    print_info("Total no. of random combinations selected in this test : {}".format(no_of_samples))
+    print_info("================================================================================")
+    #get samples not in db.
+    for db_sample in records_in_db:
+        if db_sample in all_dict:
+            all_dict.remove(db_sample)
+    if len(all_dict) == 0:
+        print_error("All the samples are tested. use --reset_results to restart test again")
```
Comment (santhosh-warrior) [inline warrior/Framework/Utils/testrandom_utils.py:194]: The error message using the old cli  option.  The option should be --reset_execution

File: warrior/Framework/Utils/testrandom_utils.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        df_fixed = df.iloc[selected_rows]
+        df = df.drop(selected_rows, axis=0)
+    #select records in df which are not in db_df
+    db_df = pd.DataFrame(records_in_db)
+    if db_df.columns.tolist():
+        df = df.merge(db_df, how = 'outer' ,indicator=True).\
+                loc[lambda x : x['_merge']=='left_only']
+        df = df.drop(['_merge'], axis = 1)
+    if no_of_samples and len(df.index) == 0:
+        print_error("All the samples are tested. use --reset_results to restart test again")
```
Comment (santhosh-warrior) [inline warrior/Framework/Utils/testrandom_utils.py:234]: Use --reset_execution

**Pull request merged**